SwanProjects/src/ProjectDialog.tsx

```diff
@@ -0,0 +1,218 @@
+// Copyright (c) SWAN Team.
+// Author: Omar Zapata CERN 2021
+
+import { showErrorMessage } from '@jupyterlab/apputils';
+import { showDialog } from './dialog';
+import {
+  contentRequest,
+  createProjectRequest,
+  editProjectRequest
+} from './request';
+import { Spinner } from '@jupyterlab/apputils';
+import { CommandRegistry } from '@lumino/commands';
+/**
+ * Namespace for project dialogs
+ */
+export namespace ProjectDialog {
+  export interface ISWANOptions {
+    name?: string;
+    stack?: string;
+    release?: string;
+    platform?: string;
+    user_script?: string;
+    corrupted?: boolean;
+    stacks_options?: { [stack: string]: { [release: string]: Array<string> } };
+  }
+
+  /**
+   * variables from backend response
+   */
+  export interface ISWANReqResponse {
+    status: boolean;
+    project_dir: string;
+    msg: string;
+  }
+
+  /**
+   * Create and show a modal dialog to create or modify projects.
+   *
+   * @param options - The dialog setup options.
+   * @param create - true for a new project, false to modify.
+   * @param commands - CommandRegistry object
+   * @param theme - colors in the interface 'light' | 'dark'.
+   * @returns A promise that resolves with whether the dialog was accepted
+   */
+  // eslint-disable-next-line  no-inner-declarations
+  export async function OpenModal(
+    options: ISWANOptions,
+    create: boolean,
+    commands: CommandRegistry,
+    theme: 'light' | 'dark'
+  ): Promise<any> {
+    const _spinner = new Spinner();
+    const old_options = Object.assign({}, options);
+
+    function startSpinner(): void {
+      /**
+       * Function to start the spiner in the SwanLauncer, embed in the html tag with id jp-main-dock-panel.
+       */
+      const node = document.getElementById('jp-main-dock-panel');
+      node?.appendChild(_spinner.node);
+      node?.focus();
+      _spinner.activate();
+      _spinner.show();
+      _spinner.node.focus();
+    }
+
+    /**
+     * hides the spiner from the component
+     */
+    function stopSpinner(): void {
+      _spinner.hide();
+    }
+
+    let valid = false;
+    let dialogResult: {
+      changesSaved: boolean;
+      newOptions?: ISWANOptions;
+    } | null = null;
+    do {
+      dialogResult = await showDialog({
+        ...options,
+        theme
+      });
+      if (dialogResult?.changesSaved && dialogResult?.newOptions) {
+        options = dialogResult.newOptions;
+        if (options.name?.trim() !== '') {
+          //check is project already exists
+          if (create) {
+            const content = await contentRequest(
+              'SWAN_projects/' + options.name
```

Comment:
To be changed when we allow projects anywhere.

Reply:
Yes, of course, It has to be modified when we support projects anywhere.